src/CacheWithResilience.php

```diff
@@ -0,0 +1,165 @@
+            // set $returnStaleValue to true
+            if (false === $returnStaleValue && new DateTime() > $value['expiresAt']) {
+                return new CacheItem();
+            }
+        }
+
+        // in setItem() value gets wrapped into an array. set the item to its original value
+        $val = $item->get();
+        $item->set($val['value']);
+
+        return $item;
+    }
+
+    /**
+     * @param CacheItemInterface $item
+     * @param mixed $value
+     * @param int|string|DateTimeInterface $ttl
+     *   TTL in seconds, or a constant from CacheInterface, or a DateTime to expire at
+     * @return bool
+     */
+    public function setItem(CacheItemInterface $item, $value, $ttl): bool
+    {
+        // this condition calculates the DateTime when the element should be consider as expire
+        if ($ttl instanceof DateTimeInterface) {
+            $expiresAt = $ttl;
+        } elseif (\in_array($ttl, [0, -1, CacheInterface::NONE, CacheInterface::INDEFINITE, CacheInterface::SHORT], true)) {
+            // this case handles a list of exception that shouldn't be cached for a different length
+            $ttl = $this->calculateTtl($ttl);
+            $item->expiresAfter($ttl);
+            // getItem() expects the item to be wrapped as [value, ttl]
+            $item->set(['value' => $value, 'expiresAt' => $ttl]);
+            return $this->cachePool->save($item);
+        } else {
+            $ttl = $this->calculateTtl($ttl);
+            $expiresAt = new DateTime();
+            $expiresAt = $expiresAt->add(new DateInterval('PT' . $ttl . 'S'));
+        }
+
+        // save the expire time within the stored item so we can know if a returned key is expired.
+        $item->set(['value' => $value, 'expiresAt' => $expiresAt]);
+
+        // instead of setting the item to expire at $expiresAt we use "RESILIENCE_TTL"
+        // this way in case of error we can return the stale value
+        $item->expiresAfter(self::RESILIENCE_TTL);
+
+        return $this->cachePool->save($item);
+    }
+
+    /**
+     * IF CALLABLE RETURNS SOMETHING THAT EVALUATES TO EMPTY THE RESULT WILL NOT BE CACHED UNLESS $nullTtl IS SET
+     * TO A VALUE DIFFERENT FROM CacheInterface::NONE
+     *
+     * @param string $key
+     * @param int|string $ttl
+     *   TTL in seconds, or a constant from CacheInterface
+     * @param callable $function
+     * @param array $arguments
+     * @return mixed
+     */
+    public function getOrSet(string $key, $ttl, callable $function, array $arguments = [], $nullTtl = CacheInterface::NONE)
+    {
+        $cacheItem = $this->getItem($key);
+
+        if ($cacheItem->isHit()) {
+            return $cacheItem->get();
+        }
+
+        try {
+            $result = $function(...$arguments);
+        } catch (\Exception $e) {
+            // if exception is whitelisted, return the stale value
+            if ($this->isWhitelistedException($e)) {
+                $this->logger->error($e->getMessage());
```

Comment:
Do you think we might need more and/or different error handling in different places? 

Even if not, is this going to be enough to enable us to properly monitor when something falls back to the stale cache? We probably at least want something in the logs indicating stale items are being returned so we can alert on it.

Reply:
What about adding this line? 
```
$this->logger->warning('stale-if-error served for: ' . $cacheItem->getKey());
```
Do you have any other suggestion? Do you have any concerns about writing too many logs? I think it should be fine

Reply:
definitely add that, yeah, we can set up a metric filter and alarm for that log line